File: packages/sdk-typescript/tests/unit/recipes.test.ts
// AI-B4 SDK + doc-132 §5.2 — unit tests for RecipesResource.
//
// Pins the wire contract: method/path/body shape per endpoint, plus
// URL-encoding on id parameters so labels with spaces or slashes
// don't break route matching.

import { describe, expect, it } from 'vitest';
import { RecipesResource } from '../../src/resources/recipes.js';
import type { HttpClient } from '../../src/http.js';

interface RecordedRequest {
  method: string;
  path: string;
  body?: unknown;
  query?: Record<string, unknown>;
}

function makeFakeHttp<T>(reply: T): { http: HttpClient; calls: RecordedRequest[] } {
  const calls: RecordedRequest[] = [];
  const http = {
    request: <R>(opts: {
      method: string;
      path: string;
      body?: unknown;
      query?: Record<string, unknown>;
    }) => {
      const recorded: RecordedRequest = { method: opts.method, path: opts.path };
      if (opts.body !== undefined) recorded.body = opts.body;
      // Recorded only when NON-EMPTY, which is the property under test rather
      // than a convenience: `list()` with no arguments must put no key on the
      // wire at all, and that reads here as the absence of `query`. Recording an
      // empty object instead would make the absent case indistinguishable from
      // `{ limit: undefined }` — the defect where a client serialises the literal
      // string "undefined" into the URL. A key that exists with an undefined
      // value still has length 1, so it is still recorded and still caught.
      if (opts.query !== undefined && Object.keys(opts.query).length > 0) {
        recorded.query = opts.query;
      }
      calls.push(recorded);
      return Promise.resolve(reply as unknown as R);
    },
  } as unknown as HttpClient;
  return { http, calls };
}

describe('RecipesResource', () => {
  it('create POSTs /v1/recipes with the body verbatim', async () => {
    const reply = {
      id: 'rec_1',
      account_id: 'acc_1',
      agent_session_id: 'agt_1',
      label: 'my recipe',
      description: null,
      intent_count: 3,
      created_at: '2026-07-01T00:00:00Z',
      updated_at: '2026-07-01T00:00:00Z',
    };
    const { http, calls } = makeFakeHttp(reply);
    const recipes = new RecipesResource(http);
    const result = await recipes.create({ agent_session_id: 'agt_1', label: 'my recipe' });
    expect(calls).toEqual([
      {
        method: 'POST',
        path: '/v1/recipes',
        body: { agent_session_id: 'agt_1', label: 'my recipe' },
      },
    ]);
    expect(result).toEqual(reply);
  });

  it('get URL-encodes the id', async () => {
    const { http, calls } = makeFakeHttp({});
    await new RecipesResource(http).get('rec/with slash');
    expect(calls).toEqual([{ method: 'GET', path: '/v1/recipes/rec%2Fwith%20slash' }]);
  });

  it('delete URL-encodes the id', async () => {
    const { http, calls } = makeFakeHttp(undefined);
    await new RecipesResource(http).delete('rec/with slash');
    expect(calls).toEqual([{ method: 'DELETE', path: '/v1/recipes/rec%2Fwith%20slash' }]);
  });

  // The arm below used to be titled "list forwards pagination query params only
  // when present" while calling list() with no arguments — it asserted the ABSENT
  // direction alone, and the recorder it asserted against did not capture `query`
  // at all, so the forwarding half was untestable by construction. A dropped
  // spread in RecipesResource.list would not have reddened anything here.
  it('list with no arguments puts NO query on the wire', async () => {
    const { http, calls } = makeFakeHttp({ data: [], has_more: false, next_cursor: null });
    await new RecipesResource(http).list();
    expect(calls).toEqual([{ method: 'GET', path: '/v1/recipes' }]);
  });

  it('CRITICAL list forwards limit ALONE without inventing a cursor key', async () => {
    const { http, calls } = makeFakeHttp({ data: [], has_more: false, next_cursor: null });
    await new RecipesResource(http).list({ limit: 25 });
    expect(calls[0]?.query).toEqual({ limit: 25 });
  });

  it('CRITICAL list forwards cursor ALONE without inventing a limit key', async () => {
    const { http, calls } = makeFakeHttp({ data: [], has_more: false, next_cursor: null });
    await new RecipesResource(http).list({ cursor: 'cur_2' });
    expect(calls[0]?.query).toEqual({ cursor: 'cur_2' });
  });

  it('list forwards both params when both are given', async () => {
    const { http, calls } = makeFakeHttp({ data: [], has_more: false, next_cursor: null });
    await new RecipesResource(http).list({ limit: 10, cursor: 'cur_9' });
    expect(calls[0]?.query).toEqual({ limit: 10, cursor: 'cur_9' });
  });

  it('CRITICAL iterate threads next_cursor into the following page request', async () => {
    const page = (id: string, next: string | null): unknown => ({
      data: [
        {
          id,
          account_id: 'acc_1',
          agent_session_id: 'agt_1',
          label: 'r',
          description: null,
          intent_count: 1,
          created_at: '2026-07-01T00:00:00Z',
          updated_at: '2026-07-01T00:00:00Z',
        },
      ],
      has_more: next !== null,
      next_cursor: next,
    });
    const pages = [page('rec_1', 'cur_2'), page('rec_2', null)];
    const seen: Array<Record<string, unknown> | undefined> = [];
    let i = 0;
    const http = {
      request: (opts: { query?: Record<string, unknown> }) => {
        seen.push(opts.query);
        const r = pages[i];
        i += 1;
        return Promise.resolve(r);
      },
    } as unknown as HttpClient;

    const out: string[] = [];
    for await (const r of new RecipesResource(http).iterate({ limit: 1 })) {
      out.push((r as { id: string }).id);
    }

    expect(out, 'every page is yielded').toEqual(['rec_1', 'rec_2']);
    // The load-bearing half: page 2 must carry the cursor page 1 returned. If the
    // cursor spread were dropped, page 2 would repeat page 1's query and the walk
    // would either loop or silently stop — both invisible to a length check alone.
    expect(seen, 'page 2 carries page 1 next_cursor').toEqual([
      { limit: 1 },
      { limit: 1, cursor: 'cur_2' },
    ]);
  });

  it('suggest GETs /v1/agent-sessions/:id/recipe-suggestion with a URL-encoded id, returns the suggestion verbatim', async () => {
    const reply = {
      suggested_label: 'Fill form on example.com',
      suggested_description: 'Navigates to example.com, fills 1 field.',
      intent_count: 4,
    };
    const { http, calls } = makeFakeHttp(reply);
    const recipes = new RecipesResource(http);
    const result = await recipes.suggest('agt/with space');
    expect(calls).toEqual([
      { method: 'GET', path: '/v1/agent-sessions/agt%2Fwith%20space/recipe-suggestion' },
    ]);
    expect(result).toEqual(reply);
  });
});

File: packages/sdk-go/recipes_test.go
package driftstack

import (
	"context"
	"encoding/json"
	"net/http"
	"testing"
)

var recipeEnvelope = map[string]any{
	"id":               "rec_aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
	"account_id":       "acc_1",
	"agent_session_id": "agt_inmem_00000001",
	"label":            "login flow snapshot",
	"description":      nil,
	"intent_count":     3,
	"created_at":       "2026-05-17T12:00:00Z",
	"updated_at":       "2026-05-17T12:00:00Z",
}

func TestRecipes_Create_LabelOnly(t *testing.T) {
	t.Parallel()
	var captured map[string]any
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/recipes" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		if err := json.NewDecoder(r.Body).Decode(&captured); err != nil {
			t.Fatal(err)
		}
		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(recipeEnvelope)
	})
	got, err := client.Recipes.Create(context.Background(), CreateRecipeRequest{
		AgentSessionID: "agt_inmem_00000001",
		Label:          "login flow snapshot",
	})
	if err != nil {
		t.Fatal(err)
	}
	if got.ID != "rec_aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee" {
		t.Errorf("id=%q", got.ID)
	}
	if got.IntentCount != 3 {
		t.Errorf("intent_count=%d", got.IntentCount)
	}
	// description omitted when not supplied — clean wire shape.
	if _, present := captured["description"]; present {
		t.Errorf("description was sent on the wire when omitted from the request")
	}
}

func TestRecipes_Create_WithDescription(t *testing.T) {
	t.Parallel()
	var captured map[string]any
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if err := json.NewDecoder(r.Body).Decode(&captured); err != nil {
			t.Fatal(err)
		}
		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(recipeEnvelope)
	})
	_, err := client.Recipes.Create(context.Background(), CreateRecipeRequest{
		AgentSessionID: "agt_inmem_00000001",
		Label:          "login flow",
		Description:    "Logs into the test account.",
	})
	if err != nil {
		t.Fatal(err)
	}
	if captured["description"] != "Logs into the test account." {
		t.Errorf("description=%v", captured["description"])
	}
}

func TestRecipes_Create_NullableAgentSessionIDInResponse(t *testing.T) {
	t.Parallel()
	envelopeNullSession := make(map[string]any, len(recipeEnvelope))
	for k, v := range recipeEnvelope {
		envelopeNullSession[k] = v
	}
	envelopeNullSession["agent_session_id"] = nil
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(envelopeNullSession)
	})
	got, err := client.Recipes.Create(context.Background(), CreateRecipeRequest{
		AgentSessionID: "agt_inmem_00000001",
		Label:          "x",
	})
	if err != nil {
		t.Fatal(err)
	}
	if got.AgentSessionID != nil {
		t.Errorf("agent_session_id should decode as nil pointer when server returns null; got %v", *got.AgentSessionID)
	}
}

func TestRecipes_Suggest_URLEncodesTheSessionID(t *testing.T) {
	t.Parallel()
	suggestionEnvelope := map[string]any{
		"suggested_label":       "Fill form on example.com",
		"suggested_description": "Navigates to example.com, fills 1 field.",
		"intent_count":          4,
	}
	var gotPath string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		// r.URL.Path is the DECODED path (%2F would show as a literal
		// slash); EscapedPath() preserves the wire encoding — the same
		// pattern client_test.go / crypto_orders_test.go use to assert
		// url.PathEscape actually encoded '/' and ' '.
		gotPath = r.URL.EscapedPath()
		if r.Method != "GET" {
			t.Errorf("unexpected method: %s", r.Method)
		}
		w.WriteHeader(http.StatusOK)
		_ = json.NewEncoder(w).Encode(suggestionEnvelope)
	})
	got, err := client.Recipes.Suggest(context.Background(), "agt/with space")
	if err != nil {
		t.Fatal(err)
	}
	if gotPath != "/v1/agent-sessions/agt%2Fwith%20space/recipe-suggestion" {
		t.Errorf("path=%q", gotPath)
	}
	if got.SuggestedLabel != "Fill form on example.com" {
		t.Errorf("suggested_label=%q", got.SuggestedLabel)
	}
	if got.IntentCount != 4 {
		t.Errorf("intent_count=%d", got.IntentCount)
	}
}

// Pagination. List and Iterate had no arms at all: the tests above cover Create
// and Suggest only, so nothing asserted that a caller's Limit or Cursor ever
// reached the wire. The TS SDK carried the identical gap (V-1974), and the
// cross-SDK pagination contract is exactly the kind that drifts silently when
// only one language pins it.
func TestRecipes_List_ForwardsPaginationOnlyWhenSet(t *testing.T) {
	t.Parallel()
	var rawQuery string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/recipes" || r.Method != "GET" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		rawQuery = r.URL.RawQuery
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(RecipesListPage{Data: []Recipe{}, HasMore: false})
	})

	// The zero query: Go treats Limit 0 and Cursor "" as unset, so NOTHING may
	// reach the URL. A stray "limit=0" would be a different request, and the
	// server rejects a zero limit.
	if _, err := client.Recipes.List(context.Background(), nil); err != nil {
		t.Fatal(err)
	}
	if rawQuery != "" {
		t.Errorf("nil query should send no params, got %q", rawQuery)
	}

	// Limit alone must not invent a cursor key.
	if _, err := client.Recipes.List(context.Background(), &ListRecipesQuery{Limit: 25}); err != nil {
		t.Fatal(err)
	}
	if rawQuery != "limit=25" {
		t.Errorf("limit alone: got %q, want %q", rawQuery, "limit=25")
	}

	// Cursor alone must not invent a limit key.
	if _, err := client.Recipes.List(context.Background(), &ListRecipesQuery{Cursor: "cur_2"}); err != nil {
		t.Fatal(err)
	}
	if rawQuery != "cursor=cur_2" {
		t.Errorf("cursor alone: got %q, want %q", rawQuery, "cursor=cur_2")
	}
}

// The load-bearing one: page 2 must carry the cursor page 1 returned. Drop that
// handoff and the walk either repeats page 1 forever or stops early — neither is
// visible to a count of yielded items alone.
func TestRecipes_Iterate_ThreadsNextCursor(t *testing.T) {
	t.Parallel()
	var seenCursor []string
	cur2 := "cur_2"
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/recipes" || r.Method != "GET" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		seenCursor = append(seenCursor, r.URL.Query().Get("cursor"))
		w.Header().Set("content-type", "application/json")
		if r.URL.Query().Get("cursor") == "" {
			_ = json.NewEncoder(w).Encode(RecipesListPage{
				Data:       []Recipe{{ID: "rec_1", AccountID: "acc_1", Label: "a"}},
				HasMore:    true,
				NextCursor: &cur2,
			})
			return
		}
		_ = json.NewEncoder(w).Encode(RecipesListPage{
			Data:       []Recipe{{ID: "rec_2", AccountID: "acc_1", Label: "b"}},
			HasMore:    false,
			NextCursor: nil,
		})
	})

	var ids []string
	err := client.Recipes.Iterate(
		context.Background(),
		&ListRecipesQuery{Limit: 1},
		func(r *Recipe) (bool, error) {
			ids = append(ids, r.ID)
			return true, nil
		},
	)
	if err != nil {
		t.Fatal(err)
	}
	if len(ids) != 2 || ids[0] != "rec_1" || ids[1] != "rec_2" {
		t.Errorf("ids=%v", ids)
	}
	// Page 1 sends no cursor; page 2 must send the one page 1 returned.
	if len(seenCursor) != 2 || seenCursor[0] != "" || seenCursor[1] != "cur_2" {
		t.Errorf("cursor handoff: got %v, want [\"\" \"cur_2\"]", seenCursor)
	}
}
